internal/rollout/disruption_allowance/reconciler.go

```diff
@@ -0,0 +1,182 @@
+package disruption_allowance
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strings"
+	"time"
+
+	api "github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/internal/store"
+	"github.com/flightctl/flightctl/internal/store/selector"
+	"github.com/flightctl/flightctl/internal/tasks"
+	"github.com/flightctl/flightctl/internal/util"
+	"github.com/google/uuid"
+	"github.com/samber/lo"
+	"github.com/sirupsen/logrus"
+)
+
+const DisruptionAllowanceReconcilationInterval = 30 * time.Second
+
+type Reconciler interface {
+	Reconcile(ctx context.Context)
+}
+
+type reconciler struct {
+	store           store.Store
+	log             logrus.FieldLogger
+	callbackManager tasks.CallbackManager
+}
+
+func NewReconciler(store store.Store, callbackManager tasks.CallbackManager, log logrus.FieldLogger) Reconciler {
+	return &reconciler{
+		store:           store,
+		log:             log,
+		callbackManager: callbackManager,
+	}
+}
+
+func (r *reconciler) getFleetCounts(ctx context.Context, orgId uuid.UUID, fleet *api.Fleet) (map[string]*Counts, error) {
+	groupBy := lo.FromPtr(fleet.Spec.RolloutPolicy.DisruptionAllowance.GroupBy)
+
+	totalCounts, err := r.store.Device().CountByLabels(ctx, orgId, store.ListParams{
+		FieldSelector: selector.NewFieldSelectorFromMapOrDie(map[string]string{"metadata.owner": util.ResourceOwner(api.FleetKind, lo.FromPtr(fleet.Metadata.Name))}),
+	}, groupBy, false)
+	if err != nil {
+		return nil, err
+	}
+	annotations := lo.FromPtr(fleet.Metadata.Annotations)
+	if annotations == nil {
+		return nil, fmt.Errorf("annotations don't exist")
+	}
+
+	busyCounts, err := r.store.Device().CountByLabels(ctx, orgId, store.ListParams{
+		FieldSelector: selector.NewFieldSelectorFromMapOrDie(map[string]string{"metadata.owner": util.ResourceOwner(api.FleetKind, lo.FromPtr(fleet.Metadata.Name))}),
+	}, groupBy, true)
+	if err != nil {
+		return nil, err
+	}
+	ret, err := mergeDeviceAllowanceCounts(totalCounts, busyCounts, groupBy)
+	if err != nil {
+		return nil, err
+	}
+	return ret, nil
```

Comment:
minor: since you dont add any additional ctx to error could 

```
return mergeDeviceAllowanceCounts(totalCounts, busyCounts, groupBy)
```

Reply:
OK